mono/metadata/socket-io.c

```diff
@@ -643,6 +643,205 @@ get_family_hint (MonoError *error)
+		WSASetLastError (0);
+		return WSA_INFINITE;
+	}
+	if (timeout == 0)
+		timeout = WSA_INFINITE; // 0 means infinite
+	return timeout;
+}
+
+/*
+* Performs an alertable wait for the specified event (FD_ACCEPT_BIT,
+* FD_CONNECT_BIT, FD_READ_BIT, FD_WRITE_BIT) on the specified socket.
+* Returns TRUE if the event is fired without errors. Calls WSASetLastError()
+* with WSAEINTR and returns FALSE if the thread is alerted. If the event is
+* fired but with an error WSASetLastError() is called to set the error and the
+* function returns FALSE.
+*/
+static gboolean alertable_socket_wait (SOCKET sock, int event_bit)
+{
+	static char *EVENT_NAMES[] = {"FD_READ", "FD_WRITE", NULL /*FD_OOB*/, "FD_ACCEPT", "FD_CONNECT", "FD_CLOSE"};
+	gboolean success = FALSE;
+	int error = -1;
+	DWORD timeout = WSA_INFINITE;
+	if (event_bit == FD_READ_BIT || event_bit == FD_WRITE_BIT) {
+		timeout = get_socket_timeout (sock, event_bit == FD_READ_BIT ? SO_RCVTIMEO : SO_SNDTIMEO);
+	}
+	WSASetLastError (0);
+	WSAEVENT event = WSACreateEvent ();
+	if (event != WSA_INVALID_EVENT) {
+		if (WSAEventSelect (sock, event, (1 << event_bit) | FD_CLOSE) != SOCKET_ERROR) {
+			LOGDEBUG (g_message ("%06d - Calling WSAWaitForMultipleEvents () on socket %d", GetCurrentThreadId (), sock));
+			DWORD ret = WSAWaitForMultipleEvents (1, &event, TRUE, timeout, TRUE);
+			if (ret == WSA_WAIT_IO_COMPLETION) {
+				LOGDEBUG ( g_message ("%06d - WSAWaitForMultipleEvents () returned WSA_WAIT_IO_COMPLETION for socket %d", GetCurrentThreadId (), sock));
+				error = WSAEINTR;
+			} else if (ret == WSA_WAIT_TIMEOUT) {
+				error = WSAETIMEDOUT;
+			} else {
+				g_assert (ret == WSA_WAIT_EVENT_0);
+				WSANETWORKEVENTS ne = { 0 };
+				if (WSAEnumNetworkEvents (sock, event, &ne) != SOCKET_ERROR) {
+					if (ne.lNetworkEvents & (1 << event_bit) && ne.iErrorCode[event_bit]) {
+						LOGDEBUG (g_message ("%06d - %s error %d on socket %d", GetCurrentThreadId (), EVENT_NAMES[event_bit], ne.iErrorCode[event_bit], sock));
+						error = ne.iErrorCode[event_bit];
+					} else if (ne.lNetworkEvents & FD_CLOSE_BIT && ne.iErrorCode[FD_CLOSE_BIT]) {
+						LOGDEBUG (g_message ("%06d - FD_CLOSE error %d on socket %d", GetCurrentThreadId (), ne.iErrorCode[FD_CLOSE_BIT], sock));
+						error = ne.iErrorCode[FD_CLOSE_BIT];
+					} else {
+						LOGDEBUG (g_message ("%06d - WSAEnumNetworkEvents () finished successfully on socket %d", GetCurrentThreadId (), sock));
+						success = TRUE;
+						error = 0;
+					}
+				}
+			}
+			WSAEventSelect (sock, NULL, 0);
+		}
+		WSACloseEvent (event);
+	}
+	if (error != -1) {
+		WSASetLastError (error);
+	}
+	return success;
+}
+
+#define ALERTABLE_SOCKET_CALL(event_bit, blocking, repeat, ret, op, sock, ...) \
+	LOGDEBUG (g_message ("%06d - Performing %s " #op " () on socket %d", GetCurrentThreadId (), blocking ? "blocking" : "non-blocking", sock)); \
+	if (blocking) { \
+		if (set_blocking(sock, FALSE)) { \
```

Comment:
Are you specifically referring to the `set_blocking()` call? Would you please elaborate? If a socket is in blocking mode and a blocking call is being made it would be weird to do async calls on the same socket from other threads? Or is that supposed to be supported? Or maybe I'm misunderstanding?